File: app/src/main/java/pers/victor/androiddemos/view/MarqueeView.java
package pers.victor.androiddemos.view;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import pers.victor.androiddemos.R;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;

/**
 * Created by victor on 16-7-14
 */
public class MarqueeView extends ViewGroup implements View.OnClickListener {
    private List<String> data = new ArrayList<>();
    private OnItemClickListener onItemClickListener;
    private int tempPos = 1;
    private boolean temp = true;
    private Timer timer;
    private TimerTask timerTask;
    private int height, width;

    public MarqueeView(Context context, AttributeSet attrs) {
        super(context, attrs);
        setOnClickListener(this);
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    private void addContentViews() {
        if (data.size() == 0) {
            return;
        }
        if (data.size() == 1) {
            addTextView(0);
            return;
        }
        addTextView(0);
        addTextView(1);
    }

    private void addTextView(int index) {
        TextView textView = new TextView(getContext());
        textView.setText(data.get(index));
        textView.setMaxLines(1);
        textView.setTextSize(20);
        textView.setGravity(Gravity.CENTER);
        textView.setTextColor(ContextCompat.getColor(getContext(), R.color.white1));
        addView(textView, width, height);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        width = MeasureSpec.getSize(widthMeasureSpec);
        height = MeasureSpec.getSize(heightMeasureSpec);
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        Method method;
        int temp = 0;
        for (int i = 0; i < getChildCount(); i++) {
            getChildAt(i).layout(0, temp, width, temp + height);
            try {
                method = View.class.getDeclaredMethod("setMeasuredDimension", int.class, int.class);
                method.setAccessible(true);
                method.invoke(getChildAt(i), width, height);
            } catch (Exception e) {
                e.printStackTrace();
            }
            temp += height;
        }
    }

    public void setContentData(final List<String> data) {
        onDestroyView();
        this.data = data;
        timer = new Timer();
        timerTask = new TimerTask() {
            @Override
            public void run() {
                startTranslation();
            }
        };
        removeAllViews();
        addContentViews();
        timer.schedule(timerTask, 2000, 2000);
    }

    public void onDestroyView() {
        if (timer != null) {
            timer.purge();
            timer.cancel();
            timer = null;
        }
        if (timerTask != null) {
            timerTask.cancel();
        }
    }

    private void startTranslation() {
        for (int i = 0; i < 2; i++) {
            final ObjectAnimator animator = ObjectAnimator.ofFloat(getChildAt(temp ? i : 1 - i), "y", height * i, -height + height * i);
            animator.setDuration(500);
            if (i == 0) {
                final int finalI = i;
                animator.addListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        tempPos++;
                        if (tempPos >= data.size())
                            tempPos = 0;
                        ((TextView) getChildAt(temp ? 1 - finalI : finalI)).setText(data.get(tempPos));
                    }
                });
            }

            post(new Runnable() {
                @Override
                public void run() {
                    animator.start();
                }
            });
        }
        temp = !temp;
    }

    @Override
    public void onClick(View v) {
        if (onItemClickListener != null) {
            onItemClickListener.onItemClick((tempPos + data.size() - 1) % data.size());
        }
    }

    public interface OnItemClickListener {
        void onItemClick(int pos);
    }
}

File: app/src/main/kotlin/demos/view/MarqueeView.kt
package demos.view

import android.animation.Animator
import android.animation.ObjectAnimator
import android.content.Context
import android.support.v4.content.ContextCompat
import android.text.TextUtils
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.victor.androiddemos.R
import java.lang.reflect.Method
import java.util.*

/**
 * Created by victor on 16-7-14.
 */
class MarqueeView(context: Context, attributeSet: AttributeSet) : ViewGroup(context, attributeSet), View.OnClickListener {
    private var temp = true
    private var data: List<String> = arrayListOf()
    private var onClick: ((Int) -> Unit)? = null
    private var tempPos = 1
    private lateinit var method: Method
    private lateinit var timer: Timer
    private lateinit var timerTask: TimerTask

    override fun onClick(v: View?) {
        onClick?.invoke((tempPos + data.size - 1) % data.size)
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        method = View::class.java.getDeclaredMethod("setMeasuredDimension", Int::class.java, Int::class.java)
        method.isAccessible = true
        setOnClickListener(this)
        timer = Timer()
        timerTask = object : TimerTask() {
            override fun run() {
                startTranslation()
            }
        }
    }

    private fun addContentViews() {
        if (data.size == 0) {
            return
        }
        if (data.size == 1) {
            addTextView(0)
            return
        }
        for (i in 0..1) {
            addTextView(i)
        }
    }

    private fun addTextView(index: Int) {
        val textView = TextView(context)
        method.invoke(textView, width, height)
        textView.text = data[index]
        textView.maxLines = 1
        textView.ellipsize = TextUtils.TruncateAt.END
        textView.setPadding(60, 0, 60, 0)
        textView.gravity = Gravity.CENTER_VERTICAL
        textView.setTextColor(ContextCompat.getColor(context, R.color.blue2))
        addView(textView)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        if (!changed)
            return
        var temp = 0
        for (i in 0..childCount - 1) {
            getChildAt(i).layout(0, temp, width, temp + height)
            temp += height
        }
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        if (w > 0) {
            removeAllViews()
            addContentViews()
            timer.schedule(timerTask, 2000, 3800)
        }
    }

    fun setContentData(data: List<String>) {
        this.data = data
    }

    fun setOnItemClickListener(onClick: ((Int) -> Unit)) {
        this.onClick = onClick
    }

    private fun startTranslation() {
        for (i in 0..1) {
            val animator = ObjectAnimator.ofFloat(getChildAt(if (temp) i else 1 - i), "y", height.toFloat() * i, -height + height.toFloat() * i)
            animator.duration = 800
            if (i == 0) {
                animator.addListener(object : Animator.AnimatorListener {
                    override fun onAnimationRepeat(animation: Animator?) {
                    }

                    override fun onAnimationEnd(animation: Animator?) {
                        tempPos++
                        if (tempPos >= data.size)
                            tempPos = 0
                        (getChildAt(if (temp) 1 - i else i) as TextView).text = data[tempPos]
                    }

                    override fun onAnimationCancel(animation: Animator?) {
                    }

                    override fun onAnimationStart(animation: Animator?) {
                    }

                })
            }
            post { animator.start() }
        }
        temp = !temp
    }

}